hail/src/main/scala/is/hail/annotations/ChunkCache.scala

```diff
@@ -0,0 +1,155 @@
+  private[this] val bigChunkCache = new TreeMap[Long, LongArrayBuilder]()
+  private[this] val chunksEncountered = mutable.Map[Long, Long]()
+  private[this] val minSpaceRequirements = .9
+  private[this] var chunksRequested = 0
+  private[this] var cacheHits = 0
+  private[this] var smallChunkCacheSize = 0
+  private[this] val smallChunkCache = new Array[LongArrayBuilder](highestSmallChunkPowerOf2 + 1)
+  (0 until highestSmallChunkPowerOf2 + 1).foreach(index => {
+    smallChunkCache(index) = new LongArrayBuilder()
+  })
+
+  def getChunkSize(chunkPointer: Long): Long =  chunksEncountered(chunkPointer)
+
+  def freeChunkFromMemory(pool: RegionPool, chunkPointer: Long):Unit = {
+    val size = chunksEncountered(chunkPointer)
+    pool.incrementAllocatedBytes(-1 * size)
+    freer(chunkPointer)
+    chunksEncountered -= chunkPointer
+  }
+
+  def freeChunksFromCacheToFit(pool: RegionPool, sizeToFit: Long): Unit = {
+    var smallChunkIndex = highestSmallChunkPowerOf2
+    while((sizeToFit + pool.getTotalAllocatedBytes) > pool.getHighestTotalUsage &&
+      smallChunkIndex >= 0 && !chunksEncountered.isEmpty) {
+      if (!bigChunkCache.isEmpty) {
+        val toFree = bigChunkCache.lastEntry()
+        freeChunkFromMemory(pool, toFree.getValue.pop())
+        if (toFree.getValue.size == 0) bigChunkCache.remove(toFree.getKey)
+      }
+      else {
+        if (smallChunkCacheSize == 0) smallChunkIndex = -1
+        else {
+          val toFree = smallChunkCache(smallChunkIndex)
+          if (toFree.size != 0) {
+            freeChunkFromMemory(pool, toFree.pop())
+            smallChunkCacheSize -= 1
+          }
+          if (toFree.size == 0) smallChunkIndex -= 1
+        }
+      }
+    }
+  }
+
+  def newChunk(pool: RegionPool, size: Long): Long = {
+    if ((size + pool.getTotalAllocatedBytes) > pool.getHighestTotalUsage) {
+      freeChunksFromCacheToFit(pool, size)
+    }
+    val newChunkPointer = allocator(size)
+    chunksEncountered += (newChunkPointer -> size)
+    pool.incrementAllocatedBytes(size)
+    newChunkPointer
+  }
+
+  def freeAll(pool: RegionPool): Unit = {
+    if (!chunksEncountered.isEmpty) {
+      smallChunkCache.foreach(ab => {
+        while (ab.size > 0) {
+          freeChunkFromMemory(pool, ab.pop())
+          smallChunkCacheSize -= 1
+        }})
+      //BiConsumer needed to work with scala 2.11.12
+      bigChunkCache.forEach(new BiConsumer[Long, LongArrayBuilder]() {
+        def accept(key: Long, value: LongArrayBuilder): Unit =
+          while (value.size > 0) freeChunkFromMemory(pool, value.pop())
+      })
+    }
+  }
+
+  def getUsage(): (Int, Int) = {
+    (chunksRequested, cacheHits)
+  }
+
+  def indexInSmallChunkCache(size: Long): Int = {
+    var closestPower = highestSmallChunkPowerOf2
+    while((size >> closestPower) != 1) closestPower = closestPower - 1
+    if (size % (1 << closestPower) != 0) closestPower +=1
+    closestPower
+  }
```

Comment:
This is maybe overkill, but I think I would assert size > 0 here. If we somehow get a size of 0 here, you'll have an infinite loop. It doesn't really make sense to free something that's 0 bytes, but it's a bug that's happened before and I'd rather we get an error than an infinite loop.